clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/IngestionBatchProcessor.java

```diff
@@ -71,20 +72,36 @@ public IngestionBatchProcessor(
     this.version = Version.parseVersionFromKafkaTopicName(storeVersionName);
   }
 
+  // For testing
+  KeyLevelLocksManager getLockManager() {
+    return this.lockManager;
+  }
+
   /**
    * When {@link #lockManager} is not null, this function will try to lock all the keys
    * (except Control Messages) passed by the params.
    */
   public List<ReentrantLock> lockKeys(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records) {
     if (lockManager != null) {
       List<ReentrantLock> locks = new ArrayList<>(records.size());
+
+      /**
+       * Collect all the keys and sort them to avoid deadlock.
+       */
+      List<byte[]> keys = new ArrayList<>(records.size());
```

Comment:
Should it be a Set instead? Do you want to dedupe the keys in the batch? What if the same key is there twice and you attempt to acquire the lock for it twice? Can this scenario be unit tested?

If deduping is desired, using a Set is not the only way to achieve it of course. It could also be done as part of the latter iteration, after the list is sorted...

Reply:
I think that the current use-case dedupes the messages in the function call immediately before, so it's not "currently" an issue but could be if these locks are used elsewhere ([link](https://github.com/linkedin/venice/blob/073e61d672734234b0c434f266c541cd2b69a388/clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java#L1147))?

Reply:
That is a good question.
I think both List and Set can work as the same lock can be `locked`/`unlocked` multiple times.

I will add a unit test for repetitive keys in the same batch.

Reply:
If the deduping is elsewhere then no point in doing it again here...

That being said, I would encourage looking at the algorithm end-to-end, across all of the steps, and trying to see how we can minimize the collection traversals and instantiation and copying. Maybe there are ways to collapse multiple steps into one and do deduping and sorting at the same time (e.g. with a `TreeMap`, perhaps). I'm not prescribing to do that specifically, but just to think about it, since ingestion is sensitive to efficiency.